Ladybird: Add Autocomplete to `LocationEdit`

cammo1123 opened the pull request:
This adds a more familiar style of searching across Ladybird. No need to specify <kbd>?</kbd> before searching. URLs, are now checked against the open [public_suffix_list.dat](https://publicsuffix.org/list/public_suffix_list.datl) and if they arent found to be valid we fall through and use the default search engine.

---

Currently, a draft because This covers a lot of the codebase that I'm unfamiliar with, and id really appreciate some overall comments about things like:
 * PublicSuffixData generator is currently in `Libraries/Userland/LibWeb/PublicSuffix`. I'm not sure where quite to put this. But it is a somewhat logical place.
 * Is it okay to use the `QNetworkAccessManager` like this in ladybird?
 * Have I covered all valid URL's? (Definitely not, but I cant think of many more)

Todo:
- [ ] Fixup settings in SerenityOS 
- [ ] Find a better place for `ErrorOr<String> absolute_url(StringView url)`
- [ ] Match the available list of engines on both platforms

![image](https://github.com/SerenityOS/serenity/assets/36214028/f98e729e-45ad-498c-bb45-8e4d199510c2)

ADKaster commented:
One comment before looking at this, according to the public suffix list, it seems they would rather we only pull it "once per day" to avoid load on their servers. Our CI would end up pulling it many more times than that per day, let alone developers using it.

> The copy on publicsuffix.org, linked below, is updated daily from Github. If you wish to make your app download an updated list periodically, please use this URL and have your app download the list no more than once per day. (The list usually changes a few times per week; more frequent downloading is pointless and hammers our servers.)

cammo1123 commented:
> One comment before looking at this, according to the public suffix list, it seems they would rather we only pull it "once per day" to avoid load on their servers. Our CI would end up pulling it many more times than that per day, let alone developers using it.

I did catch that. I don't know what would be a good solution to this. They do have a copy on GitHub that we could [here](https://raw.githubusercontent.com/publicsuffix/list/master/public_suffix_list.dat). Surely GitHub could take that load?

cammo1123 commented:
Changes since the last review
* use GitHub hosted mirror so we don't overload the public suffix list requests of pulling once per day
* Added bypass PublicSuffixChecks when CTRL is pressed in Ladybird

networkException commented:
In light of https://github.com/SerenityOS/serenity/blob/58b1d9c319fbafac8318bffbf41380e1c9ee1175/Userland/Libraries/LibGUI/TextBox.cpp#L185-L186 we might want to move the suffix list data into some other library to avoid linking LibWeb into LibGUI

cammo1123 commented:
> In light of
> 
> https://github.com/SerenityOS/serenity/blob/58b1d9c319fbafac8318bffbf41380e1c9ee1175/Userland/Libraries/LibGUI/TextBox.cpp#L185-L186
> 
> we might want to move the suffix list data into some other library to avoid linking LibWeb into LibGUI

I added that FIXME When I added address highlighting, personally, I feel like the system we have now is more than adequate (Making the FIXME redundant) and that adding more granular highlighting is unnecessary. But I agree it isn't a good idea to have to link against libweb to see if a URL has a public suffix. Do you have any idea what soft of libraries would better suit it?

networkException commented:
> that adding more granular highlighting is unnecessary

I think it is very much necessary, especially in a browser but for the sake of consistency also in the whole system. I think its important to correctly reflect the internals of the engine and expose that information to the user

cammo1123 commented:
> much necessary, especially in a browser but for the sake of consistency also in the whole system. I think its important to correctly reflect the internals of the engine and expose that in

I wasn't going to remove the FIXME, if someone else wants to add it they can by all means :^). What's your suggestion on where to relocate the PublicSuffix stuff?

networkException commented:
I can only think of a LibPublicSuffix, but I'd ask on discord about this

cammo1123 commented:
Ready for review!!

Changes:
* New Library `LibPublicSuffix`
* Use a trie to search for matching domains
* Dont allow single top-level domains eg `com` or `org`

networkException commented:
Is there UI to opt out of this?

cammo1123 commented:
> Is there UI to opt out of this?

There is now :)

AtkinsSJ commented on Ladybird/Settings.h:
```diff
@@ -7,19 +7,30 @@
 
 class Settings : public QObject {
-public:
+protected:
     Settings();
```
Nit-pick: We generally order things as `public`, `protected`, then `private`.

AtkinsSJ commented on Meta/Lagom/Tools/CodeGenerators/LibPublicSuffix/GeneratePublicSuffixData.cpp:
```diff
@@ -0,0 +1,186 @@
+    ErrorOr<Optional<String>> get_public_suffix(StringView string);
+
+private:
+    Trie<char, DeprecatedString> m_dictionary;
```
Storing these as DeprecatedStrings in new code is a little gross. How about storing Strings, constructing the PublicSuffixData in a fallible factory function, and then making `the()` return ErrorOr?

AtkinsSJ commented on Meta/Lagom/Tools/CodeGenerators/LibPublicSuffix/GeneratePublicSuffixData.cpp:
```diff
@@ -0,0 +1,186 @@
+    : m_dictionary('/', "")
+{
+    // FIXME: Reduce the depth of this trie
+    for (auto str : s_public_suffixes) {
+        auto view = str.split_view("."sv);
+        view.reverse();
+
+        StringBuilder builder;
+        builder.join("."sv, view);
+        auto val = builder.string_view();
+
+        auto it = val.begin();
+        MUST(m_dictionary.insert(it, val.end(), val, [](auto& parent, auto& it) -> Optional<DeprecatedString> { 
+            return DeprecatedString::formatted("{}{}", parent.metadata_value(), *it);
+        }));
+    }
```
Might be an option to do this rearrangement loop in the generator, so that it doesn't have to happen at runtime?

AtkinsSJ reviewed (changes requested):
This doesn't currently build because of very recent changes to the URL class.

Generally looks fine. The Qt-specific things I only skimmed over though.

cammo1123 commented:
> This doesn't currently build because of very recent changes to the URL class.
> 
> Generally looks fine. The Qt-specific things I only skimmed over though.

Fixed the build issue if you want to test it out, will look into constructing a failable singleton in the morning when I've had some sleep :^)

AtkinsSJ commented:
This has a *lot* of conflicts. :sweat_smile:

cammo1123 commented:
> This has a _lot_ of conflicts. 😅

What conflicts?? (Fixed them all :^) )

AtkinsSJ commented:
This has conflicts again because the Qt files got moved. :sweat_smile:

cammo1123 commented:
> This has conflicts again because the Qt files got moved. 😅

Pretty simple fix! should be ready for review now!!

AtkinsSJ reviewed (changes requested):
The tiniest of nit-picks, otherwise I think we can merge this

cammo1123 commented:
All fixed!!

AtkinsSJ commented:
Let's goooo!
